design/20230302.gomod.md

```diff
@@ -0,0 +1,334 @@
+
+Essentially, the `test/` directory has both _actual tests_ and _test utility code_. The actual tests import several
+areas of cert-manager which become external modules under these proposals, and the utility code is imported by the core
+cert-manager module.
+
+#### Solution: Split Test Code
+
+Since there are two types of code in `test/`, we can split it.
+
+There are [known external importers](https://pkg.go.dev/github.com/cert-manager/cert-manager@v1.11.0/test/unit/gen?tab=importedby)
+of `test/unit/` which means it's difficult to move that without breaking people.
+
+As such, we could move test/e2e and test/integration - or we could make them both independent modules and keep them
+where they are.
+
+The diff for the main repository `go.mod` after separating out the tests is presented in footnote [2].
+
+### Increased Time to Patch Everything
+
+Having multiple go.mod files wil mean that when we share a dependency across many components (such as the Kubernetes
+libraries) we'll have to update multiple files rather than just one. Alternatively, if we update a dependency for the
+core `go.mod` file we'll maybe want to also update every other go.mod which imports that one.
+
+## Other Considered Approaches
+
+### Being Less Conservative
+
+The main issue we face with upgrading older versions of cert-manager is that we self-impose strict conservatism when
+it comes to any kind of backport. In this view, any change for any reason is inherently seen as bad and to be avoided,
+even if that change has no runtime impact for users.
+
+We don't need to do this. While we wouldn't seek to make major version upgrades in backports just for the fun of it,
+we could choose to accept a larger subset of backports and rely on our tests to confirm that the change is sound.
+
+This doesn't solve the problems of allowing independent control over the dependencies of different binaries, though,
+and doesn't reduce the attack surface of any of our components.
+
+### Aggressively Reducing Dependencies
+
+Rather than isolating dependencies, we could remove them by e.g. vendoring subsets of their code into our repo. This
+gives us a huge amount of control and allows us to preserve backwards compatibility very easily.
+
+It also creates a huge burden for us to maintain that vendored code, which is a drawback. We'd still have to track
+e.g. Helm to see if there are any relevant vulnerabilities reported, and then we'd have to go and actually fix them
+ourselves. If upstream code diverged significantly we might be left on our own trying to work out how to fix bugs - or
+even trying to work out if we even have a bug.
+
+There's probably some low hanging fruit we could pick here, but we're unlikely to be able to fully remove a big chunk
+of our dependencies. That means the problem won't go away - and there's always the chance that we need to add new
+dependencies down the road.
+
+## Stretch Goal: `pkg/apis` Module
```

Comment:
FWIW, as an _external consumer_ (i.e. _not_ working within the cert-manager project itself), this is pretty much all I care about

Reply:
Sure, yeah! We sorta have documented this too: https://cert-manager.io/docs/contributing/importing/#when-you-might-import-cert-manager

There have always been a few places where we've supported or expected people to import the module. This design will at least reduce the number of transitive dependencies external importers have to evaluate when importing those modules!